service/history/historyEngine.go

```diff
@@ -158,14 +157,91 @@ func (e *historyEngineImpl) StartWorkflowExecution(startRequest *h.StartWorkflow
 		return nil, &workflow.BadRequestError{Message: "Missing or invalid TaskStartToCloseTimeoutSeconds."}
 	}
 
-	// We generate a new workflow execution run_id on each StartWorkflowExecution call.  This generated run_id is
-	// returned back to the caller as the response to StartWorkflowExecution.
-	runID := uuid.New()
-	workflowExecution := workflow.WorkflowExecution{
-		WorkflowId: common.StringPtr(executionID),
-		RunId:      common.StringPtr(runID),
+	// TODO
+	execution := workflow.WorkflowExecution{
+		WorkflowId: common.StringPtr(*request.WorkflowId),
+	}
+	isBrandNewWorkflow := true
+	// without setting the run ID so we can get the current workflow. if any
+	context, release, err := e.historyCache.getOrCreateWorkflowExecution(domainID, execution)
+	execution.RunId = common.StringPtr(uuid.New())
+
+	if err != nil {
+		if _, ok := err.(*workflow.EntityNotExistsError); !ok {
+			return nil, err
+		}
+		// here we know that there is no existing run of this workflow ID, i.e. not running, no history.
+	} else {
```

Comment:
Refactor this method to not require such a long else statement.

Reply:
ok